File: apps/radio-station/server/services/weights.ts
import type {Weights} from '../types.js';

export const DEFAULT_WEIGHTS: Weights = {
    hitGravity: 70,
    goldLeak: 15,
    temperature: 40,
    separation: 60,
    powerOrbitMin: 90,
};

function clampInt(value: unknown, min: number, max: number, fallback: number): number {
    const n = typeof value === 'number' ? value : Number(value);
    if (!Number.isFinite(n)) return fallback;
    return Math.min(max, Math.max(min, Math.round(n)));
}

export function canonicalizeWeights(raw: unknown): Weights {
    const o = raw && typeof raw === 'object' ? (raw as Record<string, unknown>) : {};
    return {
        hitGravity: clampInt(o.hitGravity, 0, 100, DEFAULT_WEIGHTS.hitGravity),
        goldLeak: clampInt(o.goldLeak, 0, 100, DEFAULT_WEIGHTS.goldLeak),
        temperature: clampInt(o.temperature, 0, 100, DEFAULT_WEIGHTS.temperature),
        separation: clampInt(o.separation, 0, 100, DEFAULT_WEIGHTS.separation),
        powerOrbitMin: clampInt(o.powerOrbitMin, 60, 150, DEFAULT_WEIGHTS.powerOrbitMin),
    };
}

export function weightsJson(weights: Weights): string {
    return JSON.stringify({
        hitGravity: weights.hitGravity,
        goldLeak: weights.goldLeak,
        temperature: weights.temperature,
        separation: weights.separation,
        powerOrbitMin: weights.powerOrbitMin,
    });
}

File: apps/radio-station/api/src/main/kotlin/radioapi/domain/Weights.kt
package radioapi.domain

/** Sliders run 0..100; the orbit slider is minutes and has its own range. */
private const val PERCENT_MIN = 0
private const val PERCENT_MAX = 100
private const val POWER_ORBIT_MIN_MINUTES = 60
private const val POWER_ORBIT_MAX_MINUTES = 150

val DEFAULT_WEIGHTS: Weights = Weights(
    hitGravity = 70,
    goldLeak = 15,
    temperature = 40,
    separation = 60,
    powerOrbitMin = 90,
)

fun canonicalizeWeights(raw: Map<String, Any?>?): Weights {
    val values = raw ?: emptyMap()
    return Weights(
        hitGravity = clampInt(values["hitGravity"], PERCENT_MIN, PERCENT_MAX, DEFAULT_WEIGHTS.hitGravity),
        goldLeak = clampInt(values["goldLeak"], PERCENT_MIN, PERCENT_MAX, DEFAULT_WEIGHTS.goldLeak),
        temperature = clampInt(values["temperature"], PERCENT_MIN, PERCENT_MAX, DEFAULT_WEIGHTS.temperature),
        separation = clampInt(values["separation"], PERCENT_MIN, PERCENT_MAX, DEFAULT_WEIGHTS.separation),
        powerOrbitMin = clampInt(
            values["powerOrbitMin"],
            POWER_ORBIT_MIN_MINUTES,
            POWER_ORBIT_MAX_MINUTES,
            DEFAULT_WEIGHTS.powerOrbitMin,
        ),
    )
}

fun weightsJson(weights: Weights): String = "{\"hitGravity\":${weights.hitGravity},\"goldLeak\":${weights.goldLeak}," +
    "\"temperature\":${weights.temperature},\"separation\":${weights.separation}," +
    "\"powerOrbitMin\":${weights.powerOrbitMin}}"

private fun clampInt(value: Any?, min: Int, max: Int, fallback: Int): Int {
    val number = when (value) {
        is Number -> value.toDouble()
        is String -> value.toDoubleOrNull()
        else -> null
    } ?: return fallback
    if (!number.isFinite()) return fallback
    return kotlin.math.round(number).toInt().coerceIn(min, max)
}
